doc/developer/design/20210629_source_trait.md

```diff
@@ -0,0 +1,319 @@
+between many `Instance`s on workers of the same node. For this reason, each
+`SourceConfig` implementation can define a `State` type that will contain any
+shared resources between instances. The indended use of this type is to
+instanciate a singleton per-node and use it any time a source is instantiated.
+
+It is crucial that any state stored in `State` be considered transient and safe
+to lose at any point since it is not persisted in the catalog.
+
+* **Source SQL syntax is simple enough to derive directly type definitions**
+
+This is more of an assumption for the future rather than a statement about the
+current sources since if we look at the existing ones we can find violations of
+it. The goal of this assumption is to strongly urge new source implementations
+to rely on just their struct definition and a `Deserialize` implementation for
+their parsing needs instead of parsing directly. The benefit is twofold. Firstly,
+it is much easier to define the expected syntax, and secondly our syntax will
+naturally follow a pattern as a consequence of not giving arbitrary access to
+the parser.
+
+## Source architecture
+
+The following diagram describes the state of the system after a `CREATE
+MATERIALIZED SOURCE foo ...` statement. Here is what happened.
+
+When the user typed the statement, the system used `Source::Sql` to parse the
+options specified. Then, the system constructed a `Source` using
+`Source::from_sql`, which had the opportunity to validate the parameters. Since
+`Source` implements `ToSql`, the created instance is serialized and stored in
+the catalog in order to re-create it on restarts.
+
+Since the user requested a materialized view the system begins to instanciate
+the source. To do that, the coordinator called `Source::create_instances(4)` to
+signal to the source that there are 4 workers available. This particular
+source has only three shards, and so it returned three `Config`s to the
+coordinator. These `Config` objects contain the necessary information to be
+converted into Instances once they reach the dataflow layer. `Config` are
+serialized and stored in the catalog so that cleanup can be performed once the
+source is `DROP`ed. Ultimately, these `Config` objects are packed into a
+`DataflowDesc` and sent to the dataflow layer for instantiation.
+
+In this example, the dataflow layer consists of two nodes with two theads each.
+Since there were only three `Config` objects in the received `DataflowDesc` the
+first node is assigned the first two, and the second one the third. This
+assignment is arbitrary and can be based on more elaborate logic to balance the
+work out.
+
+The nodes are now ready to create `Instance`s from each `Config` and start
+receiving data into the dataflow. They do this using `Config::into_instance()`,
+which requires a mutable reference to a `State`. Each dataflow node constructs,
+and keeps around, a `State` object that it re-uses across all instantiations of
+a `Config`.
```

Comment:
I will adjust the wording. In reality, it shouldn't matter if the state is per-process, per-thread, or if the implementation does `config.into_instance(&mut State::default()` every time. We can definitely move this in the workers.

One detail though, when we talked one of the examples you brought up was shared libraries. I'm not very familiar with them but I think these are loaded process-wide. It's not a huge problem since that particular `State` could have a static variable inside it in case it is instantiated multiple times.